### ISSUE TITLE
Failed to attach disk 'LocalState\ext4.vhdx' to WSL2: The system cannot find the file specified.

### ISSUE THREAD

**Issue by liudonghua123** [microsoft/WSL#10032]
### Windows Version

Microsoft Windows [Version 10.0.22621.1555]

### WSL Version

1.2.5.0

### Are you using WSL 1 or WSL 2?

- [X] WSL 2
- [ ] WSL 1

### Kernel Version

Linux version 5.15.90.1-microsoft-standard-WSL2 (oe-user@oe-host) (x86_64-msft-linux-gcc (GCC) 9.3.0, GNU ld (GNU Binutils) 2.34.0.20200220) #1 SMP Fri Jan 27 02:56:13 UTC 2023

### Distro Version

Kali linux

### Repro Steps

1. uninstall the old kali linux
2. reinstall kali linux from ms store.

### Expected Behavior

The Kali linux installed successfully.

### Actual Behavior

Could not open/start it when installed finished.

### Diagnostic Logs

```
Failed to attach disk 'C:\Users\Liu.D.H\AppData\Local\Packages\KaliLinux.54290C8133FEE_ey8k8hqnwqnmg\LocalState\ext4.vhdx' to WSL2: The system cannot find the file specified.
Error code: Wsl/Service/CreateInstance/MountVhd/ERROR_FILE_NOT_FOUND
Press any key to continue...
```

**Comment by cakeGH** [comment]
Hello, I'm new to this computer stuff but I had this issue too. Couldn't find anything except a short 2min YT video. Basically you need to re-install and run your Unix installation through Powershell first. 

Steps:
run Windows Powershell as Admin
wsl --uninstall kali-linux
wsl --install --distribution kali-linux

Not sure how or why it works but it worked for me. Hope this helps you and whoever encounters this!

**Comment by davhdavh** [comment]
`>wsl --unregister ubuntu`
followed by
`>wsl --install`
fixes the problem for ubuntu atleast

**Comment by dreamflasher** [comment]
I can confirm, a simple reinstallation via Microsoft Store does not work, but wsl unregister and then install does.

**Comment by kilogold** [comment]
> I can confirm, a simple reinstallation via Microsoft Store does not work, but wsl unregister and then install does.

Agreed. Same issue. This is the way.

**Comment by bi44fv** [comment]
> `>wsl --unregister ubuntu` followed by `>wsl --install` fixes the problem for ubuntu atleast

it worked for me, thanks alot

**Comment by githubrowdy** [comment]
> `>wsl --unregister ubuntu` followed by `>wsl --install` fixes the problem for ubuntu atleast

It does also for kali
wsl --unregsiter kali-linux

**Comment by mcscwizzy** [comment]
Can confirm the unregister and install works. Saved my backside today.

**Comment by azuraerae** [comment]
> I can confirm, a simple reinstallation via Microsoft Store does not work, but wsl unregister and then install does.

This is it 👍

**Comment by pauljones0** [comment]
This fixed it. Thank you

**Comment by V4mp1reeee** [comment]
i have it loading, its saying installing ubuntu, im worried.

**Comment by kumarchandresh** [comment]
`> wsl --unregister ubuntu`
followed by reinstall from ms store also worked.

**Comment by alizul01** [comment]
> `>wsl --unregister ubuntu` followed by `>wsl --install` fixes the problem for ubuntu atleast

worked!

**Comment by bmaciag** [comment]
> `>wsl --unregister ubuntu` followed by `>wsl --install` fixes the problem for ubuntu atleast

worked. thx

**Comment by marviniciuz** [comment]
I have the same problem how to solve?

**Comment by marviniciuz** [comment]
I achieved
after running the commands
wsl --unregister ubuntu and wsl --instal

after uninstalling ubuntu from the microsoft store and installing again

**Comment by sradu86** [comment]
in pws:
wsl --unregister kali-linux

wsl  --install kali-linux

**Comment by meyerbro** [comment]
It's kind of obvious that loosing all your work would fix it guys.
But what's the solution when we can't/don't want to lose all the data???

**Comment by TheFlash787** [comment]
> It's kind of obvious that loosing all your work would fix it guys. But what's the solution when we can't/don't want to lose all the data???

If your vhdx file is missing, it might be that your data is already gone -- as that file has been deleted which is the storage volume for the Ubuntu operating system. Uninstalling and reinstalling just reinstates that fresh volume.

**Comment by Senbonzakura1234** [comment]
Facing this problem too

**Comment by johnnymatt** [comment]
Works (as of 11/09/2023):
wsl --unregister ubuntu
wsl -- install

**Comment by daniel-rodriguez-azumo** [comment]
> It's kind of obvious that loosing all your work would fix it guys. But what's the solution when we can't/don't want to lose all the data???

### **TL;DR: Go to "Turn windows features on and off" -> uncheck windows subsystem for linux -> restart -> do it again and check the windows subsystem for linux box and after restarting again you are good to go.**

I had the same issue as OP, I tried moving my `ext4.vhdx` disk file, in an attempt to free it from whatever process was using it, it denied this action but also told me which process was using this file, it was `System` in my case, so I tried the steps described on my tldr and it worked, without losing data or any side effects.

I also tried other solutions which didn't work: Signing out of my user, restarting, closing other processes.

Hope this helps.

**Comment by Silic0nS0ldier** [comment]
I've run into the same problem. Tried using [Recuva](https://www.ccleaner.com/recuva) to see if I could restore the file with no luck.

I wonder if this data loss is related to behaviours associated with these directories. From the [UWP docs](https://learn.microsoft.com/en-us/windows/uwp/get-started/fileio-learning-track);
* Each UWP app gets several folders under ` c:\users<user name>\AppData\Local\Packages<app package identifier>\`, which appears to be the case for WSL distros as well. At least those distributed via the store.
* `LocalState` is for data local to the current device. OneDrive backs up this data, such that it can be restored on new devices.
* There is also `RoamingState` (in Windows 11 it no longer roams) and `TempState`.
Such backup behaviour should be readonly for the current device, so I do not know _how_ it could cause this. I just know from the docs that it is a mechanism that touches this area and that something abnormal has occurred.

Possibly related issues;
* No cause determined - https://stackoverflow.com/questions/52713201/downloaded-files-in-appdata-local-packages-auto-deleting-in-uwp
* Certificate issue with side-loaded app - https://stackoverflow.com/questions/74595055/uwp-storage-file-is-deleted-on-close-even-though-i-did-not-how-can-i-prevent-it\
  I know on Android if the certificate used to sign an app changes, data-loss occurs. I believe the same applies to UWP/sandboxed apps on Windows. So in my case if the certificate for Ubuntu 22.04 changed with an app update, that may have wiped the data (possibly in a manner that leaves _just enough_ behind that errors are raised rather than a new install being started).

Looking at the last modified date in MS Store, I can see "Ubuntu 22.04.2 LTS" was modified on 2023-09-03, which is between my last usage and discovery of this issue. Currently installed version is `2204.2.47.0`.

Assuming the issue was caused by `LocalState` being wiped due to certificate changes, preventing reoccurrence means having the WSL files live somewhere else. The easiest way to achieve this would be;
1. Install as normal.
2. Open shell (not in installed distro, Windows Powershell is what I used)
3. Export distro\
   `wsl --export <distro-name> $env:UserProfile\Downloads\distro.tar`
4. Delete distro\
   `wsl --unregister <distro-name>`
5. Import exported distro\
   `wsl --import <distro-name> <install-location> $env:UserProfile\Downloads\distro.tar`

----

Output is slightly different to what OP posted, so here it what I got (to aid future searches).
```
Failed to attach disk 'C:\Users\jorda\AppData\Local\Packages\CanonicalGroupLimited.Ubuntu22.04LTS_79rhkp1fndgsc\LocalState\ext4.vhdx' to WSL2: The system cannot find the file specified.
Error code: Wsl/Service/CreateInstance/MountVhd/ERROR_FILE_NOT_FOUND

[process exited with code 4294967295 (0xffffffff)]
You can now close this terminal with Ctrl+D, or press Enter to restart.
```

**Comment by dobigbadwolf** [comment]
i had to adapt the unix command so it was 'wsl unregister kali-linux' the ran the 2nd linux command and it worked perfectly thanx!

**Comment by Reyuki-san** [comment]
I've also had a similar problem, and trying @daniel-rodriguez-azumo's solution didn't work for me. In my case, wsl error after updating [the user directory](https://github.com/coreybutler/nvm-windows/issues/1033)

what should I do? how to update the `OldPath` to `CurrentPath` on users directory?

**Comment by Reyuki-san** [comment]
Ah NVM, https://github.com/microsoft/WSL/issues/9349#issuecomment-1379447153 solved my problem

I hope in the future this process will be done automatically when updating the user directory.

**Comment by gexadax** [comment]
Thanks this is working 👍 
wsl --unregister ubuntu
wsl -- install

**Comment by TuanDung0902** [comment]
> `>wsl --unregister ubuntu` followed by `>wsl --install` fixes the problem for ubuntu atleast

Yeah, it worked!

**Comment by hamishwillee** [comment]
This seems to work for me on Garden in WSL2 (when running PX4 autopilot). I still get the error `libEGL warning: failed to open /dev/dri/renderD128: Permission denied` but the UI does launch and all the other Ogre errors have disappeared.

**Comment by TiloGit** [comment]
had this today. Got lucky just reboot of Windows fixed it. (Microsoft Windows NT 10.0.22621.0)

**Comment by xanthisafk** [comment]
> `>wsl --unregister ubuntu` followed by `>wsl --install` fixes the problem for ubuntu atleast

This is the one. Thank you!

**Comment by CihatDinc** [comment]
> `>wsl --unregister ubuntu` ardından `>wsl --install` en azından ubuntu için sorun düzeltildi
it worked, thanks :)

**Comment by gayashvihangana** [comment]
> `>wsl --unregister ubuntu` followed by `>wsl --install` fixes the problem for ubuntu atleast

This worked for me. Thank you!

**Comment by BJBUniverse** [comment]
>wsl --unregister kali-linux followed by >wsl --install kali-linux worked

**Comment by Leonezz** [comment]
> `>wsl --unregister ubuntu` followed by `>wsl --install` fixes the problem for ubuntu atleast

One should NOTE this operation removes the distro entirely

**Comment by dmtzs** [comment]
Executing the command: wsl --unregister adn then wsl --install doenst fix the problem.
What really does is to install another ubuntu which is the latest version too with LTS, and is not the same as the one in the microsoft store.

But if for a reason you prefer to install it like this way then you should be sure to delete the broken version if you dont want to have trash in your pc

**Comment by adnanh5** [comment]
> `>wsl --unregister ubuntu` followed by `>wsl --install` fixes the problem for ubuntu atleast

Thank you so much for this solution I was struggling for hours

**Comment by adnanh5** [comment]
> `>wsl --unregister ubuntu` followed by `>wsl --install` fixes the problem for ubuntu atleast

Thank you so much for this solution I was struggling for hours

**Comment by ueeou** [comment]
Thank you so much

**Comment by Smiley2507** [comment]
> `>wsl --unregister ubuntu` followed by `>wsl --install` fixes the problem for ubuntu atleast

Thank you. This solved my problem.

**Comment by safwanshamsir99** [comment]
I found another solution on [David Bombal YouTube](https://youtu.be/_fntjriRe48?si=cYD7byIDksiN51aP&t=276)

**Comment by ASV1995** [comment]
> `>wsl --unregister ubuntu` followed by `>wsl --install` fixes the problem for ubuntu atleast

Still works thank you :)

**Comment by dobigbadwolf** [comment]
Hey there,

I'm glad to hear that the fix worked for Ubuntu! Thanks for the help.

Cheers!

**Comment by fikfr** [comment]
help, i cant unregister it, cos powershell says there is no wsl installed, even tho it says there is... 
![image](https://github.com/microsoft/WSL/assets/82052838/8c3baa2f-3c9b-4214-843d-c4d7343fcbe3)
when starting kali, it still says the same error...

**Comment by Emereh-boop** [comment]
thank a  lot this fixed the issue 👍

> `>wsl --unregister ubuntu` followed by `>wsl --install` fixes the problem for ubuntu atleast

**Comment by Fabienjulio** [comment]
> Works (as of 11/09/2023): wsl --unregister ubuntu wsl -- install

Thanks a lot!
I had the same problem but with Docker Desktop
`provisioning docker WSL distros: ensuring main distro is deployed: checking if main distro is up to date: checking main distro bootstrap version: getting main distro bootstrap version: exit code: 4294967295: running WSL command wsl.exe C:\Windows\System32\wsl.exe -d docker-desktop -u root -e wsl-bootstrap version: Échec de l’attachement du disque « <HOME>\AppData\Local\Docker\wsl\distro\ext4.vhdx » à WSL2. Error code: Wsl/Service/CreateInstance/MountVhd/HCS/ERROR_FILE_NOT_FOUND : exit status 0xffffffff`

Solved it with just `wsl --unregister docker-desktop` and then restarting Docker

**Comment by jjavierdguezas** [comment]
> Thanks a lot! I had the same problem but with Docker Desktop `provisioning docker WSL distros: ensuring main distro is deployed: checking if main distro is up to date: checking main distro bootstrap version: getting main distro bootstrap version: exit code: 4294967295: running WSL command wsl.exe C:\Windows\System32\wsl.exe -d docker-desktop -u root -e wsl-bootstrap version: Échec de l’attachement du disque « <HOME>\AppData\Local\Docker\wsl\distro\ext4.vhdx » à WSL2. Error code: Wsl/Service/CreateInstance/MountVhd/HCS/ERROR_FILE_NOT_FOUND : exit status 0xffffffff`
> 
> Solved it with just `wsl --unregister docker-desktop` and then restarting Docker


same problem here, `wsl --unregister docker-desktop` and restart Docker Desktop just work! thanks @Fabienjulio ... no data was lost btw

**Comment by sayol** [comment]
> `>wsl --unregister ubuntu` followed by `>wsl --install` fixes the problem for ubuntu atleast

Confirm that this works perfectly!

**Comment by EricHerlitz** [comment]
This is usually an issue with your rights, ensure you have sufficient rights to the vhdx file. Which you probably don't have since this file most likely was moved from another computer.
E.g
![image](https://github.com/user-attachments/assets/00bb2787-33fe-4270-b487-367402148679)

**Comment by chozillla** [comment]
it is 2024 and this works for me! 

`wsl --unregister ubuntu
wsl -- install`

**Comment by omereliusuk** [comment]
thank you

**Comment by RobertoAlessandri** [comment]
> `>wsl --unregister ubuntu` followed by `>wsl --install` fixes the problem for ubuntu atleast

What if I don't want to --unregister? I am afraid I'd loose al my (non backed-up) files

**Comment by luwukien** [comment]
> `>wsl --unregister ubuntu` followed by `>wsl --install` fixes the problem for ubuntu atleast

thank you. It's working.

**Comment by hanif-akbar** [comment]
> `>wsl --unregister ubuntu`
> followed by
> `>wsl --install`
> fixes the problem for ubuntu atleast

worked!

**Comment by SnoepNFTs** [comment]
For debian, this worked for me:
`> wsl --unregister debian`
'> wsl --install -d debian'

**Comment by pvanheus** [comment]
For those doing `wsl --unregister` followed by `wsl --install` - does this not result in data loss??

**Comment by pwalessi-dell** [comment]
> For those doing `wsl --unregister` followed by `wsl --install` - does this not result in data loss??

Yes. It wiped my data in wsl.

**Comment by pf-f** [comment]
why you call this solution, is there any other workaround?  i do not want to lose my data

**Comment by AustinBrunkhorst** [comment]
> > For those doing `wsl --unregister` followed by `wsl --install` - does this not result in data loss??
> 
> Yes. It wiped my data in wsl.

Not only this, but it doesn't prompt any confirmation. I learned this the hard way.

**Comment by naXa777** [comment]
> is there any other workaround?

@pf-f for those who have their vhdx file and want to simply fix permissions without loosing the data, let me share a powershell script:

```ps1
# run as admin when `wsl -d Ubuntu` for non-admin current user gives an error like:
# Nie można dołączyć dysku „\\?\D:\backup\WSL\Ubuntu.vhdx” do podsystemu WSL2: Access is denied.
# Kod błędu: Wsl/Service/CreateInstance/MountDisk/HCS/E_ACCESSDENIED
# or Docker shows error like:
# WSL integration with distro 'Ubuntu' unexpectedly stopped. Do you want to restart it?
# configuring docker in Ubuntu: docker cli config: failed to write file: exit status 0xffffffff

wsl --shutdown
takeown /F "D:\backup\WSL\Ubuntu.vhdx"
$me = "$env:USERDOMAIN\$env:USERNAME"
icacls "D:\backup\WSL\Ubuntu.vhdx" /grant "${me}:(F)" /grant "ALL APPLICATION PACKAGES:(RX)" /inheritance:e

# check new permissions
icacls "D:\backup\WSL\Ubuntu.vhdx"

wsl --shutdown
```
(replace "D:\backup\WSL\Ubuntu.vhdx" with a real location)

verification:
```ps1
# confirm Ubuntu shows “Running/Stopped” and “2”
wsl -l -v
# launch the distro
wsl -d Ubuntu
```

**Comment by stew675** [comment]
> [@pf-f](https://github.com/pf-f) for those who have their vhdx file and want to simply fix permissions without loosing the data, let me share a powershell script:

Thank you.  This actually works and does not result in any data loss

**Comment by shuaiyuanxx** [comment]
We would like to better understand the root cause of this issue; however, no logs are currently available for analysis. If you encounter this issue again, we would appreciate it if you could collect and share the relevant logs, as this would help us improve code quality. Thank you.

<details>

<summary>How to collect WSL logs</summary>

Download and execute [collect-wsl-logs.ps1](https://github.com/Microsoft/WSL/blob/master/diagnostics/collect-wsl-logs.ps1) in an **administrative powershell prompt**:

```
Invoke-WebRequest -UseBasicParsing "https://raw.githubusercontent.com/microsoft/WSL/master/diagnostics/collect-wsl-logs.ps1" -OutFile collect-wsl-logs.ps1
Set-ExecutionPolicy Bypass -Scope Process -Force
.\collect-wsl-logs.ps1
```

The script will output the path of the log file once done.

If this is a networking issue, please use `.\collect-wsl-logs.ps1 -LogProfile networking` instead, following the instructions in [Collect WSL logs for networking issues](https://github.com/microsoft/WSL/blob/master/CONTRIBUTING.md#collect-wsl-logs-for-networking-issues)

Once completed please upload the output files to this GitHub issue.
 
See [Collect WSL logs (recommended method)](https://github.com/microsoft/WSL/blob/master/CONTRIBUTING.md#8-collect-wsl-logs-recommended-method).

If you choose to email these logs instead of attaching them to the bug, please send them to wsl-gh-logs@microsoft.com with the GitHub issue number in the subject, and include a link to your GitHub issue comment in the message body, and reply with '/emailed-logs'.
</details>

**Comment by xziy** [comment]
`wsl --unregister ubuntu`

This deleted the file from my machine's disk wsl, the data was lost

**Comment by keithomayot** [comment]
This might also help. I moved my vhd to a different partition to manage space and got the same error.  Tried mounting the vhd but was unsuccessful. The solution for was as simple as editing the Windows registry where the linux Distro is found as follows:
1. Open the Registry Editor and search the failed Distro. This can be found at `Computer\HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Lxss\`. You can easily locate it by searching 'Lxss' using the window menu [view -> find -> search 'Lxss'].
2. Select the distro with error found listed in this section.
3. Edit the BasePath value to the new folder where the vhd file is located. (NOTE: Make sure to Unmount the vhd from wsl if you had mounted it).
And every thing should work as expected. This works without having to unregister the Distro.
